# üèóÔ∏è Regras de Valida√ß√£o Estrutural do Censo Escolar

## üìã Vis√£o Geral

As **Regras Estruturais** s√£o diferentes das **Regras de Registro**. Enquanto as regras de registro validam campos individuais dentro de cada registro, as regras estruturais validam:

- Estrutura geral do arquivo
- Sequ√™ncia de registros
- Relacionamentos entre diferentes registros
- Limites e restri√ß√µes globais
- Codifica√ß√£o e caracteres

## üèóÔ∏è Estrutura de Arquivos

```
src/validation/rules/
‚îú‚îÄ‚îÄ base-structural.rule.ts                    # Classe base abstrata para regras estruturais
‚îú‚îÄ‚îÄ structural-rules-manager.service.ts        # Gerenciador de regras estruturais
‚îî‚îÄ‚îÄ structural-rules/
    ‚îú‚îÄ‚îÄ file-structure.rule.ts                 # Valida√ß√£o da estrutura geral do arquivo
    ‚îú‚îÄ‚îÄ record-sequence.rule.ts                # Valida√ß√£o da sequ√™ncia de registros
    ‚îú‚îÄ‚îÄ school-structure.rule.ts               # Valida√ß√£o da estrutura individual de escolas
    ‚îú‚îÄ‚îÄ character-encoding.rule.ts             # Valida√ß√£o de caracteres e codifica√ß√£o
    ‚îî‚îÄ‚îÄ README.md                              # Esta documenta√ß√£o
```

## üìö Regras Implementadas

### 1. **FileStructureRule** - Estrutura Geral do Arquivo

Valida aspectos globais do arquivo:

- ‚úÖ Presen√ßa do registro 99 (fim de arquivo)
- ‚úÖ N√∫mero de campos correto por tipo de registro (00, 10, 20, 30, 40, 50, 60)
- ‚úÖ Limite m√°ximo de 100 escolas por arquivo
- ‚úÖ Presen√ßa de pelo menos uma escola

**Campos esperados por registro:**

- Registro 00: 56 campos
- Registro 10: 187 campos
- Registro 20: 70 campos
- Registro 30: 108 campos
- Registro 40: 7 campos
- Registro 50: 38 campos
- Registro 60: 32 campos

### 2. **RecordSequenceRule** - Sequ√™ncia de Registros

Valida a ordem correta dos registros conforme especifica√ß√£o do Censo:

- **Registro 00** ‚Üí Deve vir no in√≠cio ou ap√≥s registros 40 ou 60
- **Registro 10** ‚Üí Deve vir ap√≥s registro 00
- **Registro 20** ‚Üí Deve vir ap√≥s registros 10 ou 20
- **Registro 30** ‚Üí Deve vir ap√≥s registros 00, 20 ou 30
- **Registro 40** ‚Üí Deve vir ap√≥s registros 30 ou 40
- **Registro 50** ‚Üí Deve vir ap√≥s registros 40 ou 50
- **Registro 60** ‚Üí Deve vir ap√≥s registros 50 ou 60

### 3. **SchoolStructureRule** - Estrutura Individual de Escolas

Valida a estrutura de cada escola conforme sua situa√ß√£o de funcionamento:

#### **Escola Em Atividade (situa√ß√£o = 1):**

- ‚úÖ Deve ter registros: 00, 10, 20, 30, 40
- ‚úÖ Todas as turmas (20) devem ter alunos (60)
- ‚úÖ Todas as turmas (20) devem ter profissionais (50)
- ‚úÖ M√°ximo de 3 gestores (registro 40)
- ‚úÖ M√°ximo de 1.500 turmas

#### **Escola Paralisada/Extinta (situa√ß√£o = 2 ou 3):**

- ‚úÖ Deve ter apenas registros: 00, 30, 40
- ‚ùå N√£o deve ter registros: 10, 20, 50, 60

#### **Valida√ß√µes de Relacionamento:**

- Se h√° alunos (60), deve haver turmas (20)
- Se h√° profissionais (50), deve haver turmas (20)
- Cada turma deve ter pelo menos um aluno
- Cada turma deve ter pelo menos um profissional

### 4. **CharacterEncodingRule** - Caracteres e Codifica√ß√£o

Valida caracteres permitidos e codifica√ß√£o do arquivo:

#### **Caracteres Permitidos:**

- ‚úÖ Letras MAI√öSCULAS (A-Z)
- ‚úÖ N√∫meros (0-9)
- ‚úÖ Caracteres especiais b√°sicos
- ‚ùå Letras min√∫sculas (a-z)
- ‚ùå Acentos e caracteres especiais latinos

#### **Codifica√ß√£o:**

- ‚úÖ Deve ser ISO-8859-1 (caracteres 0x00 a 0xFF)
- ‚ùå N√£o deve ter BOM (Byte Order Mark)
- ‚ùå N√£o deve ter caracteres Unicode fora do padr√£o ISO-8859-1
- ‚ö†Ô∏è Aviso sobre mistura de terminadores de linha (CRLF vs LF)

## üéØ Como Funciona

### Fluxo de Valida√ß√£o Estrutural:

```typescript
// 1. O StructuralValidatorService delega para o gerenciador
const errors = structuralValidatorService.validateStructure(records);

// 2. O gerenciador executa todas as regras na ordem correta
StructuralRulesManagerService
  ‚îú‚îÄ> FileStructureRule          // Valida estrutura b√°sica
  ‚îú‚îÄ> CharacterEncodingRule      // Valida caracteres
  ‚îú‚îÄ> RecordSequenceRule         // Valida sequ√™ncia (se estrutura OK)
  ‚îî‚îÄ> SchoolStructureRule        // Valida escolas individuais (se estrutura OK)
```

### Contexto de Valida√ß√£o:

As regras estruturais trabalham com um **contexto compartilhado**:

```typescript
interface StructuralValidationContext {
  records: string[]; // Todos os registros do arquivo
  schoolStructures: Map<string, SchoolStructure>; // Estruturas de cada escola
  totalSchools: number; // Total de escolas no arquivo
  hasRecord99: boolean; // Se tem registro de fim
  fileContent?: string; // Conte√∫do completo (para valida√ß√£o de encoding)
}
```

## üîß Como Usar

### **Uso B√°sico:**

```typescript
import { StructuralValidatorService } from './validators/structural-validator.service';

// Validar estrutura completa
const errors = await structuralValidator.validateStructure(records);

// Validar apenas caracteres e codifica√ß√£o
const charErrors = await structuralValidator.validateCharacters(fileContent);
const encodingErrors = await structuralValidator.validateEncoding(fileContent);
```

### **Uso Granular (atrav√©s do gerenciador):**

```typescript
import { StructuralRulesManagerService } from './rules/structural-rules-manager.service';

// Validar apenas estrutura do arquivo
const fileErrors = await manager.validateFileStructureOnly(records);

// Validar apenas sequ√™ncia de registros
const seqErrors = await manager.validateRecordSequenceOnly(records);

// Validar apenas estruturas das escolas
const schoolErrors = await manager.validateSchoolStructuresOnly(records);

// Validar apenas caracteres e codifica√ß√£o
const charErrors = await manager.validateCharactersAndEncoding(content);
```

## üÜï Como Adicionar Uma Nova Regra Estrutural

### **Passo 1: Criar a Regra**

```typescript
// src/validation/rules/structural-rules/my-structural-rule.ts
import { Injectable } from '@nestjs/common';
import {
  BaseStructuralRule,
  StructuralValidationContext,
} from '../base-structural.rule';
import { ValidationError } from '../../../common/interfaces/validation.interface';

@Injectable()
export class MyStructuralRule extends BaseStructuralRule {
  constructor() {
    super('my_rule', 'Descri√ß√£o da minha regra');
  }

  validate(context: StructuralValidationContext): ValidationError[] {
    const errors: ValidationError[] = [];

    // Implementar valida√ß√£o aqui
    // Use: context.records, context.schoolStructures, etc.

    // Criar erros usando o helper
    errors.push(
      this.createError(
        lineNumber,
        recordType,
        fieldName,
        fieldPosition,
        fieldValue,
        'rule_name',
        'Mensagem de erro',
      ),
    );

    return errors;
  }
}
```

### **Passo 2: Registrar no Gerenciador**

```typescript
// src/validation/rules/structural-rules-manager.service.ts
constructor(
  private readonly fileStructureRule: FileStructureRule,
  private readonly myStructuralRule: MyStructuralRule, // ‚Üê Adicionar
) {}

validateStructure(records: string[], fileContent?: string): ValidationError[] {
  const errors: ValidationError[] = [];
  const context = this.buildValidationContext(records, fileContent);

  // Adicionar valida√ß√£o
  const myErrors = this.myStructuralRule.validate(context);
  errors.push(...myErrors);

  return errors;
}
```

### **Passo 3: Registrar no M√≥dulo**

```typescript
// src/validation/validation.module.ts
import { MyStructuralRule } from './rules/structural-rules/my-structural-rule';

@Module({
  providers: [
    // ... outros providers
    MyStructuralRule, // ‚Üê Adicionar
  ],
})
export class ValidationModule {}
```

## üéØ Benef√≠cios da Arquitetura

### ‚úÖ **Separa√ß√£o de Responsabilidades**

- Cada regra tem uma responsabilidade espec√≠fica
- F√°cil de testar isoladamente

### ‚úÖ **Reutiliza√ß√£o de C√≥digo**

- Classe base com m√©todos auxiliares
- Contexto compartilhado entre regras

### ‚úÖ **Manutenibilidade**

- Cada regra em arquivo separado
- F√°cil de adicionar/remover regras

### ‚úÖ **Flexibilidade**

- Pode executar todas as regras ou apenas algumas
- Ordem de execu√ß√£o controlada

### ‚úÖ **Consist√™ncia**

- Mesmo padr√£o de erros em todas as regras
- Nomenclatura padronizada

## üìä Exemplos de Erros

### **Estrutura de Arquivo:**

```json
{
  "lineNumber": 0,
  "recordType": "FILE",
  "fieldName": "school_limit",
  "ruleName": "file_structure_too_many_schools",
  "errorMessage": "N√∫mero de escola por arquivo (100 escolas) excedido.",
  "severity": "error"
}
```

### **Sequ√™ncia de Registros:**

```json
{
  "lineNumber": 15,
  "recordType": "20",
  "fieldName": "record_sequence",
  "ruleName": "record_sequence_invalid_sequence",
  "errorMessage": "Registro 20 declarado em linha inadequada, ap√≥s o registro 00. Esperado ap√≥s: registro 10 ou registro 20",
  "severity": "error"
}
```

### **Estrutura de Escola:**

```json
{
  "lineNumber": 0,
  "recordType": "20",
  "fieldName": "class_without_students",
  "ruleName": "school_structure_class_needs_students",
  "errorMessage": "Turma informada sem aluno(a) vinculado a ela.",
  "severity": "error"
}
```

### **Caracteres e Codifica√ß√£o:**

```json
{
  "lineNumber": 5,
  "recordType": "00",
  "fieldName": "field_3",
  "ruleName": "character_encoding_invalid_characters",
  "errorMessage": "O campo cont√©m caractere(s) n√£o permitido(s): \"√°\".",
  "severity": "error"
}
```

## üöÄ Pr√≥ximos Passos

Poss√≠veis melhorias futuras:

1. **Valida√ß√µes Avan√ßadas:**
   - Valida√ß√£o de CPF/CNPJ
   - Valida√ß√£o de CEP
   - Valida√ß√£o de datas l√≥gicas

2. **Performance:**
   - Cache de contexto
   - Paraleliza√ß√£o de valida√ß√µes independentes

3. **Relat√≥rios:**
   - Sugest√µes de corre√ß√£o
   - Estat√≠sticas de erros

4. **Testes:**
   - Testes unit√°rios para cada regra
   - Testes de integra√ß√£o

---

**Sistema pronto para expans√£o! üéâ**
